Multi-serialiser support

iky opened the pull request:
Current Nameko setup allows for having one serialisation format across the whole cluster. Serialisation format can be set by `serializer` config entry.

This change adds support for multiple serialisation formats:
* adds support for accepting multiple serialisation formats for built-in messaging, pub-sub and RPC consumers
* RPC replies in the same format as was the request
* ads the ability to define custom serializers in config

Two new config entries added. `ACCEPT` defines serialisation formats accepted by Nameko consumers:
``` yaml
ACCEPT:
    - json
    - yaml
````

`SERIALIZERS` sets up serialisers registry. Adding a custom serialiser:
``` yaml
SERILIZERS:
    myserializer:
        encode: mymodule.encode
        decode: mymodule.decode
        content_type: application:x-mymimetype
        content_encoding: utf-8
```

An example of a service using a custom XJSON format for serialising but accepting both, the built-in JSON and the custom XJSON:
``` yaml
serializer: json
ACCEPT:
    - json
    - xjson
SERIALIZERS:
    xjson:
        encode: xjson.encode
        decode: xjsom.decode
        content_type: application:x-json
```

With the setup above, a service would:
* publish events using JSON
* handle events serialized as both JSON and XJSON
* send RPC request using JSON
* accept RPC requests as both JSON and XJSON, replying to JSON RPC calls with JSON and XJSON RPC calls with XJSON

mattbennett commented on nameko/containers.py:
```diff
@@ -168,6 +166,21 @@ def __init__(self, service_cls, config):
+            SERIALIZER_CONFIG_KEY, DEFAULT_SERIALIZER)
+
+        self.accept = self.config.get(
+            ACCEPT_CONFIG_KEY, [self.serializer])
```
I think we need to check whether the specified serializer and accept values have been registered as serializers.

iky replied:
True, it should be checked and it should raise `ConfigurationError` with a clear message rather than `KeyErrors` later on and on various places.

mattbennett commented on nameko/rpc.py:
```diff
@@ -188,13 +188,17 @@ def send_response(self, result, exc_info):
 
+        # send response encoded the same way as was the request message
+        content_type = self.message.properties['content_type']
+        serializer = kombu.serialization.registry.type_to_name[content_type]
```
Reinforcing the comment above: we should check earlier than here that there is a registered serializer for this `content_type`.

iky replied:
True, it should be checked and it should raise ConfigurationError with a clear message rather than KeyErrors later on and on various places.

A check added.

mattbennett commented on test/test_serialization.py:
```diff
@@ -208,3 +219,170 @@ def echo(self, arg):
+
+    payload = {'spam': 'ham'}
+
+    # Forwarder serialiser is set to JSON and should send and receive JSON
```
Might be worth splitting this into multiple tests. If you put them all into a class you can do the setup in fixtures to avoid duplication

iky replied:
Sure thing, parameterised.

mattbennett commented on test/test_serialization.py:
```diff
@@ -208,3 +219,170 @@ def echo(self, arg):
+    assert '{spam: ham}' in msg['payload']
+    assert msg['properties']['content_type'] == 'application/x-yaml'
+
+    assert called.mock_calls == [call(payload), call(payload)]
```
Same here. I think separate tests for these two cases would be more readable (and more targeted if one fails). You might even be able to parametrize them

iky replied:
Sure thing, parameterised.

iky commented:
Thanks for the review @mattbennett!

I added multi-serialization support also to standalone RPC proxy and same serializer setup to standalone event publisher too. For that I extracted the serialization setup to its own module.
